csrc/trtllm_flavored_gemm_runner.cu

```diff
@@ -0,0 +1,243 @@
+            << "Failed to get device attribute: " << cudaGetErrorString(cudaStatus);
+        cached_multi_processor_count = count;
+        cached_device_index = device_index;
+        return count;
+      }
+    }();
+
+    TVM_FFI_ICHECK(gemm.run(config, workspace, gemmData, static_cast<void*>(stream),
+                            multiProcessorCount, true, globalTrtllmGenFlavoredGemmModuleCache) == 0)
+        << "Error occurred when running flavored TRTLLM-GEN GEMM!";
+  }
+
+  std::vector<int64_t> getValidTactics(int64_t m, int64_t n, int64_t k) const {
+    auto const gemm = gemm::gemm::GemmInterface();
+    auto const configs = gemm.getGemmConfigs();
+
+    auto const gemmData = createGemmData(m, n, k);
+
+    std::vector<int64_t> validTactics{};
+    for (auto const& configIndex : mPassingConfigIndices) {
+      auto const& config = configs[configIndex];
+      if (gemm.isValidConfig(config, gemmData)) {
+        validTactics.push_back(configIndex);
+      }
+    }
+    return validTactics;
+  }
+
+ private:
+  TrtllmGenFlavoredGemmRunnerOptions mOptions;
+  std::vector<int64_t> mPassingConfigIndices;
+};
+
+using tvm::ffi::Array;
+using tvm::ffi::Optional;
+
+void trtllm_flavored_gemm(Tensor workspace_buffer, Tensor a, Tensor b, Tensor globalScale,
+                          Tensor out, int64_t tactic) {
+  if (tactic == -1) {
+    TVM_FFI_LOG_AND_THROW(RuntimeError)
+        << "Default tactic is not supported for flavored TRTLLM-GEN GEMM. Please use auto-tuning "
+           "to find the best kernel for your problem.";
+  }
+  CHECK_DEVICE(a, b);
+  CHECK_DEVICE(a, out);
+  CHECK_INPUT(a);
+  CHECK_INPUT(b);
+  CHECK_INPUT(out);
+  CHECK_INPUT(workspace_buffer);
+  CHECK_DIM(2, a);
+  TVM_FFI_ICHECK(b->ndim == 3) << "b must be a block layout matrix (3D tensor with "
+                                  "dims [N/BLOCK_SIZE, K, BLOCK_SIZE])";
+  TVM_FFI_ICHECK_EQ(a->dtype, b->dtype);
+  TVM_FFI_ICHECK(a->dtype == dl_float8_e4m3fn) << "a must be a Float8 tensor";
+
+  int32_t m = a->shape[0];
+  int32_t k = a->shape[1];
+  int32_t n = b->shape[1];
+  auto const blockSize = b->shape[2];
+  auto const kFromB = b->shape[0] * blockSize;
+  TVM_FFI_ICHECK(kFromB == a->shape[1]) << "Matrix dimensions don't match for multiplication";
+  TVM_FFI_ICHECK(out->shape[0] == m && out->shape[1] == n) << "Output tensor has wrong dimensions";
+
+  auto runner =
+      flashinfer::TrtllmGenFlavoredGemmRunner(flashinfer::TrtllmGenFlavoredGemmRunnerOptions{
+          .eltType = gemm::trtllm::gen::Dtype::E4m3,
+          .outputType = gemm::trtllm::gen::Dtype::Bfloat16,
+      });
+
+  auto stream = get_stream(a->device);
+
+  int64_t const required_workspace_size = runner.getWorkspaceSizeInBytes(m, n, k, tactic);
+  int64_t const provided_workspace_size =
+      get_numel(workspace_buffer) * get_element_size(workspace_buffer);
+  if (provided_workspace_size < required_workspace_size) {
+    TVM_FFI_LOG_AND_THROW(RuntimeError)
```

Comment:
can we internally alloc workspace instead of error out here? we can follow `csrc\trtllm_gemm_runner.cu`

Reply:
Good point. I wanted to avoid internal allocation on purpose here. The caller can appropriately size the workspace using `getWorkspaceSizeInBytes`, and be certain that the library will not allocate more. For the caller, this avoids:
- Unexpected OOMs
- Unexpected latency peaks
I know this is not always possible, but here it is and I would strongly advise for it. It makes life much easier for the inference frameworks that will depend on Flashinfer.

For the other cases, when dynamic allocations cannot be excluded, I would suggest making it explicit by having the op accept some allocator as input.

Reply:
Agreed it's a better practice to avoid internal allocation here.